api/src/wfl/environment.clj

```diff
@@ -0,0 +1,74 @@
+(ns wfl.environment
+  "Map environment to various values here."
+  (:require [clojure.data.json  :as json]
+            [clojure.tools.logging :as log]
+            [clojure.string :as str]
+            [vault.client.http]         ; vault.core needs this
+            [vault.core :as vault]
+            [wfl.util  :as util]))
+
+;; TODO: `is-known-cromwell-url?` in modules means new projects require new releases
+;;  since this is baked in code. Can we improve this?
+
+(def cromwell-label-keys
+  "Use these keys to extract Cromwell labels from workflow inputs."
+  [:data_type
+   :project
+   :regulatory_designation
+   :sample_name
+   :version])
+
+(def gotc-dev-cromwell
+  {:labels cromwell-label-keys
+   :monitoring_script "gs://broad-gotc-prod-cromwell-monitoring/monitoring.sh"
+   :url "https://cromwell-gotc-auth.gotc-dev.broadinstitute.org"})
+
+(def debug
+  "A local environment for development and debugging."
+  {:cromwell gotc-dev-cromwell
+   :data-repo
+   {:service-account "jade-k8-sa@broad-jade-dev.iam.gserviceaccount.com"
+    :url "https://jade.datarepo-dev.broadinstitute.org"}
+   :google
+   {:jes_roots ["gs://broad-gotc-dev-cromwell-execution"]
+    :projects ["broad-gotc-dev"]}
+   :server
+   {:project "broad-gotc-dev"
+    :service-account "secret/dsde/gotc/dev/wfl/wfl-non-prod-service-account.json"
+    :vault "secret/dsde/gotc/dev/zero"}})
+
+(def stuff
+  "Map ENVIRONMENT and so on to vault paths and JDBC URLs and so on."
+  (letfn [(make [m e] (let [kw  (keyword e)
+                            var (resolve (symbol e))]
+                        (-> (var-get var)
+                            (assoc :doc (:doc (meta var)) :keyword kw :name e)
+                            (->> (assoc m kw)))))]
+    (reduce make {} ["debug"])))
+
+(defn ^:private vault-secrets
+  "Return the secrets at `path` in vault."
+  [path]
+  (let [token (or (->> [(System/getProperty "user.home") ".vault-token"]
+                       (str/join "/") slurp util/do-or-nil)
+                  (System/getenv "VAULT_TOKEN"))]
+    (try (vault/read-secret
+          (doto (vault/new-client "https://clotho.broadinstitute.org:8200/")
+            (vault/authenticate! :token token))
+          path {})
+         (catch Throwable e
+           (log/warn e "Issue with Vault")
+           (log/debug "Perhaps run 'vault login' and try again")))))
+
+(def ^:private defaults
+  "Default values for environment variables, mainly for dev purposes."
+  {"GOOGLE_APPLICATION_CREDENTIALS"  (-> "secret/dsde/gotc/dev/wfl/wfl-non-prod-service-account.json"
+                                         vault-secrets
+                                         (json/write-str :escape-slash false)
+                                         .getBytes)
+   "TERRA_DATA_REPO_URL" "https://jade.datarepo-dev.broadinstitute.org/"
+   "WFL_OAUTH2_CLIENT_ID" (-> "secret/dsde/gotc/dev/zero" vault-secrets :oauth2_client_id)
```

Comment:
I didn't realise you needed to perform IO here. I don't think we should be doing this stuff when the namespace is being loaded. Perhaps hide this behind a thunk:
```clojure
(def ^:private defaults
  "Thunks for default values for environment variables, mainly for dev purposes."
  {"GOOGLE_APPLICATION_CREDENTIALS" 
   #(-> "secret/dsde/gotc/dev/wfl/wfl-non-prod-service-account.json"
        vault-secrets
        (json/write-str :escape-slash false)
        .getBytes)
   "TERRA_DATA_REPO_URL"
    (constantly "https://jade.datarepo-dev.broadinstitute.org/")
...
)
```
Then change `getenv` to apply that form:
```clojure
(def getenv (memoize (fn [name] (or (System/getenv name) ((defaults name))))))
```

Reply:
That's a really good point! I updated the code and ended up having something like `#(-> "42")` for consistency.

Reply:
Kinda have to list all of the used ENV VARs here in defaults otherwise we will get NPE. (good thing is listing all vars here indeed fulfills the original need of documenting everything together at one place )